cloudygames/views.py

```diff
@@ -5,111 +5,87 @@
-        order_by = ['name']
-        read_only_fields = ('id',)
-
-class GameSessionFilter(django_filters.FilterSet):
-    game = django_filters.CharFilter(name='game__id')
-    player = django_filters.CharFilter(name='player_username')
-
-    class Meta:
-        model = GameSession
-        fields = ['game', 'player']
-
-class PlayerSaveDataFilter(django_filters.FilterSet):
-    game = django_filters.CharFilter(name='game__id')
-    player = django_filters.CharFilter(name='player_username')
-
-    class Meta:
-        model = PlayerSaveData
-        fields = ['game', 'player']
-
-
-######################## ViewSet ############################
-
 class GameViewSet(viewsets.ModelViewSet):
     serializer_class = GameSerializer
     filter_class = GameFilter
+    permission_classes = (OperatorOnly,)
 
     def get_queryset(self):
         is_owned = self.request.query_params.get('owned', 0)
         if is_owned == '1':
-            _user = self.request.user
-            return Game.objects.filter(users=_user)
+            user = self.request.user
+            owned_games_id = GameOwnership.objects.filter(user=user).values_list('game__id', flat=True)
+            return Game.objects.filter(pk__in=owned_games_id)
         return Game.objects.all().order_by('name')
 
+class GameOwnershipViewSet(viewsets.ModelViewSet):
+    serializer_class = GameOwnershipSerializer
+    filter_class = GameOwnershipFilter
+    permission_classes = (UserIsOwnerOrOperatorExceptUpdate,)
+
+    def get_queryset(self):
+        user = self.request.user
+        if(user.is_staff):
+            return GameOwnership.objects.all()
+        return GameOwnership.objects.filter(user=user)
+
 class GameSessionViewSet(viewsets.ModelViewSet):
     serializer_class = GameSessionSerializer
     filter_class = GameSessionFilter
+    permission_classes = (UserIsOwnerOrOperatorExceptUpdate,)
     
     def get_queryset(self):
         user = self.request.user
         if(user.is_staff):
             return GameSession.objects.all()
-        return GameSession.objects.filter(player=user)
+        return GameSession.objects.filter(user=user)
 
-    def put(self, request):
+    def create(self, request):
         serializer = GameSessionSerializer(data=request.data)
 
         if serializer.is_valid():
-            user = self.request.user
-            gameid = serializer.data['game']
+            game = serializer.validated_data['game']
+            user = serializer.validated_data['user']
 
-            try:
-                game = Game.objects.get(id=gameid)
-            except Game.DoesNotExist: # Case 1: Invalid game
-                return Response(status=status.HTTP_400_BAD_REQUEST)
-
-            try:
-                session = GameSession.objects.get(player=self.request.user, game=game) # Case 2: Already joined
-            except GameSession.DoesNotExist:
+            if game.id in GameOwnership.objects.filter(user=user).values_list('game__id', flat=True): # User owns the game
```

Comment:
Nope, it will return HTTP 400 BAD REQUEST instead. I have tested this and it works =D